src/factory-reset/factory-reset-tool.c

```diff
@@ -0,0 +1,381 @@
+
+        r = efi_set_variable_string(EFI_SYSTEMD_VARIABLE_STR("FactoryResetRequest"), formatted);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set EFI variable FactoryResetRequest: %m");
+
+        log_debug("Set EFI variable FactoryResetRequest to '%s'.", formatted);
+
+        if (!arg_quiet)
+                log_info("Factory reset requested.");
+
+        return 0;
+}
+
+static int verb_cancel(int argc, char *argv[], void *userdata) {
+        int r;
+
+        FactoryResetMode f = factory_reset_mode();
+        if (f < 0)
+                return log_error_errno(f, "Failed to determine current factory reset mode: %m");
+        if (f == FACTORY_RESET_ON)
+                return log_error_errno(SYNTHETIC_ERRNO(EBUSY), "System already executing factory reset, cannot cancel.");
+        if (f != FACTORY_RESET_PENDING) {
+                log_info("No factory reset has been requested, cannot cancel, skipping.");
+                return 0;
+        }
+
+        if (!is_efi_boot()) {
+                if (!arg_quiet)
+                        log_info("Not an EFI boot, cannot remove FactoryResetMode EFI variable, not cancelling.");
+
+                return 0;
+        }
+
+        r = efi_set_variable(EFI_SYSTEMD_VARIABLE_STR("FactoryResetRequest"), /* value= */ NULL, /* size= */ 0);
+        if (r < 0)
+                return log_error_errno(r, "Failed to remove FactoryResetRequest EFI variable: %m");
+
+        if (!arg_quiet)
+                log_info("Factory reset cancelled.");
+
+        return 0;
+}
+
+static int retrigger_block_devices(void) {
+        int r;
+
+        /* Let's retrigger block devices after factory reset is complete: it's quite likely that some
+         * partitions went away or got recreated, and will only be considered relevant once factory reset
+         * mode is left. For example, /dev/disk/gpt-auto-root is like that: it is only created once factory
+         * reset mode is complete. */
+
+        if (!udev_available()) {
+                log_info("Skipping triggering of block devices, as udev is not available.");
+                return 0;
+        }
+
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        r = sd_device_enumerator_new(&e);
+        if (r < 0)
+                return log_error_errno(r, "Failed to allocate device enumerator: %m");
+
+        r = sd_device_enumerator_allow_uninitialized(e);
+        if (r < 0)
+                return log_error_errno(r, "Failed to enable enumeration of uninitialized devices: %m");
+
+        r = sd_device_enumerator_add_match_subsystem(e, "block", /* match = */ true);
+        if (r < 0)
+                return log_error_errno(r, "Failed to filter device enumeration by 'block' subsystem: %m");
+
+        if (!arg_quiet)
+                log_info("Retriggering block devices.");
+
+        FOREACH_DEVICE(e, d) {
+                r = sd_device_trigger(d, SD_DEVICE_CHANGE);
```

Comment:
It may be useful to also introduce `--settle` argument for waiting for uevents being processed.

Otherwise, even after reset-complete.service is finished, gpt-auto-root sometimes may not exist when a later service is started.

Reply:
But that's fine I think, nothing should be waiting for this synchronously. Waiting/mounting a block device always has been an asynchronous affair, hence I think that's fine here too and expected.

(I am open to changing this one day, if we actually need this for anything, but right now we here certainly do not)

Reply:
yeah, it is not necessary to do in this PR.